File: plldb-server/src/main/java/cn/twincest/plldb/dao/UserDao.java
package cn.twincest.plldb.dao;

import cn.twincest.plldb.domain.User;
import cn.twincest.plldb.util.DatabaseUtil;
import cn.twincest.plldb.util.TypeUtil;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Updates;
import com.mongodb.lang.Nullable;
import com.mongodb.lang.NonNull;
import org.bson.types.ObjectId;

import java.util.Date;
import java.util.List;

public class UserDao {

	private static final MongoCollection<User> COLLECTION = DatabaseUtil.getUserCollection();

	@NonNull
	public static List<User> findAll() {
		return DatabaseUtil.convertIterableToList(COLLECTION.find());
	}

	@Nullable
	public static User findById(
			@NonNull ObjectId id
	) {
		return COLLECTION.find(Filters.eq("_id", id)).first();
	}

	@Nullable
	public static User findByName(
			@NonNull String name
	) {
		return COLLECTION.find(Filters.eq("name", name)).first();
	}

	public static boolean hasName(
			@NonNull String name
	) {
		return findByName(name) != null;
	}

	@Nullable
	public static User register(
			@NonNull String name,
			@NonNull String rawPassword,
			@Nullable String email
	) {
		// make data
		var registerDate = new Date();
		var user = new User(null, name, TypeUtil.encryptPassword(rawPassword), email, false, false, false, registerDate, registerDate);
		// insert
		var insertOneResult = COLLECTION.insertOne(user);
		// find insert result
		var result = COLLECTION.find(Filters.eq("_id", insertOneResult.getInsertedId().asObjectId())).first();
		return result;
	}

	@Nullable
	public static User login(
			@NonNull String name,
			@NonNull String rawPassword
	) {
		// check name and password
		var filter = Filters.eq("name", name);
		var user = COLLECTION.find(filter).first();
		if (user == null) {
			return null;
		}
		if (!TypeUtil.checkPassword(rawPassword, user.getPassword())) {
			return null;
		}
		// update loginDate
		var loginDate = new Date();
		var update = Updates.set("loginDate", loginDate);
		// update
		var result = COLLECTION.findOneAndUpdate(filter, update, DatabaseUtil.DEFAULT_FIND_ONE_AND_UPDATE_OPTIONS);
		return result;
	}

	@Nullable
	public static User updatePassword(
			@NonNull ObjectId id,
			@NonNull String rawPassword
	) {
		// make rule
		var filter = Filters.eq("_id", id);
		var update = Updates.set("password", TypeUtil.encryptPassword(rawPassword));
		// update
		var result = COLLECTION.findOneAndUpdate(filter, update, DatabaseUtil.DEFAULT_FIND_ONE_AND_UPDATE_OPTIONS);
		return result;
	}

	@Nullable
	public static User updateEmail(
			@NonNull ObjectId id,
			@Nullable String email
	) {
		// make rule
		var filter = Filters.eq("_id", id);
		var update = Updates.set("email", email);
		// update
		var result = COLLECTION.findOneAndUpdate(filter, update, DatabaseUtil.DEFAULT_FIND_ONE_AND_UPDATE_OPTIONS);
		return result;
	}

	public static boolean format() {
		var all = findAll();
		var deleteResult = COLLECTION.deleteMany(Filters.empty());
		var deleteCount = deleteResult.getDeletedCount();
		System.out.println("delete count : " + deleteCount);
		var insertResult = COLLECTION.insertMany(all);
		var insertCount = insertResult.getInsertedIds().size();
		System.out.println("insert count : " + insertCount);
		return deleteCount == insertCount;
	}

}

File: plldb-server/src-refactor/main/java/cn/twin_kles/plldb/dao/UserDao.kt
package cn.twin_kles.plldb.dao

import cn.twin_kles.plldb.util.DatabaseUtil
import cn.twin_kles.plldb.domain.User
import org.bson.types.ObjectId
import com.mongodb.client.model.Filters
import cn.twin_kles.plldb.util.TypeUtil
import com.mongodb.client.model.Updates
import java.util.*

object UserDao {

    private val COLLECTION = DatabaseUtil.userCollection

    fun findAll(): List<User> {
        return DatabaseUtil.convertIterableToList(COLLECTION.find())
    }

    fun findById(id: ObjectId): User? {
        return COLLECTION.find(Filters.eq("_id", id)).first()
    }

    fun findByName(name: String): User? {
        return COLLECTION.find(Filters.eq("name", name)).first()
    }

    fun hasName(name: String): Boolean {
        return findByName(name) != null
    }

    fun register(name: String, rawPassword: String, email: String?): User? {
        // make data
        val registerDate = Date()
        val user = User(
            null,
            name, TypeUtil.encryptPassword(rawPassword),
            email,
            false, false, false,
            registerDate, registerDate
        )
        // insert
        val insertOneResult = COLLECTION.insertOne(user)
        // find insert result
        return COLLECTION.find(Filters.eq("_id", insertOneResult.insertedId.asObjectId())).first()
    }

    fun login(name: String, rawPassword: String): User? {
        // check name and password
        val filter = Filters.eq("name", name)
        val user = COLLECTION.find(filter).first() ?: return null
        if (!TypeUtil.checkPassword(rawPassword, user.password!!)) {
            return null
        }
        // update loginDate
        val loginDate = Date()
        val update = Updates.set("loginDate", loginDate)
        // update
        return COLLECTION.findOneAndUpdate(filter, update, DatabaseUtil.DEFAULT_FIND_ONE_AND_UPDATE_OPTIONS)
    }

    fun updatePassword(id: ObjectId, rawPassword: String): User? {
        // make rule
        val filter = Filters.eq("_id", id)
        val update = Updates.set("password", TypeUtil.encryptPassword(rawPassword))
        // update
        return COLLECTION.findOneAndUpdate(filter, update, DatabaseUtil.DEFAULT_FIND_ONE_AND_UPDATE_OPTIONS)
    }

    fun updateEmail(id: ObjectId, email: String?): User? {
        // make rule
        val filter = Filters.eq("_id", id)
        val update = Updates.set("email", email)
        // update
        return COLLECTION.findOneAndUpdate(filter, update, DatabaseUtil.DEFAULT_FIND_ONE_AND_UPDATE_OPTIONS)
    }

    fun format(): Boolean {
        val all = findAll()
        val deleteResult = COLLECTION.deleteMany(Filters.empty())
        val deleteCount = deleteResult.deletedCount
        println("delete count : $deleteCount")
        val insertResult = COLLECTION.insertMany(all)
        val insertCount = insertResult.insertedIds.size
        println("insert count : $insertCount")
        return deleteCount == insertCount.toLong()
    }
}